rllib/examples/metrics/custom_metrics_in_env_runners.py

```diff
@@ -0,0 +1,339 @@
+which should allow you to set breakpoints anywhere in the RLlib code and
+have the execution stop there for inspection and debugging.
+
+For logging to your WandB account, use:
+`--wandb-key=[your WandB API key] --wandb-project=[some project name]
+--wandb-run-name=[optional: WandB run name (within the defined project)]`
+
+
+Results to expect
+-----------------
+This script has not been finetuned to actually learn the environment. Its purpose
+is to show how you can create and log custom metrics during episode sampling and
+have these stats be sent to WandB for further analysis.
+
++---------------------+----------+----------------+--------+------------------+
+| Trial name          | status   | loc            |   iter |   total time (s) |
+|                     |          |                |        |                  |
+|---------------------+----------+----------------+--------+------------------+
+| PPO_env_efd16_00000 | RUNNING  | 127.0.0.1:6181 |      4 |          72.4725 |
++---------------------+----------+----------------+--------+------------------+
++------------------------+------------------------+------------------------+
+|    episode_return_mean |   num_episodes_lifetim |   num_env_steps_traine |
+|                        |                      e |             d_lifetime |
+|------------------------+------------------------+------------------------|
+|                  76.4  |                     45 |                   8053 |
++------------------------+------------------------+------------------------+
+"""
+from typing import Optional, Sequence
+
+import gymnasium as gym
+import matplotlib.pyplot as plt
+from matplotlib.colors import Normalize
+import numpy as np
+
+from ray.rllib.algorithms.callbacks import DefaultCallbacks
+from ray.rllib.env.wrappers.atari_wrappers import wrap_atari_for_new_api_stack
+from ray.rllib.utils.images import resize
+from ray.rllib.utils.test_utils import (
+    add_rllib_example_script_args,
+    run_rllib_example_script_experiment,
+)
+from ray.tune.registry import get_trainable_cls, register_env
+
+
+class MsPacmanHeatmapCallback(DefaultCallbacks):
+    """A custom callback to extract information from MsPacman and log these.
+
+    This callback logs:
+    - the positions of MsPacman over an episode to produce heatmaps from this data.
+    At each episode timestep, the current pacman (y/x)-position is determined and added
+    to the episode's temporary storage. At the end of an episode, a simple 2D heatmap
+    is created from this data and the heatmap is logged to the MetricsLogger (to be
+    viewed in WandB).
+    - the max distance travelled by MsPacman per episode, then averaging these max
+    values over a window of size=100.
+    - the mean distance travelled by MsPacman per episode (over an infinite window).
+    - the number of lifes of MsPacman EMA-smoothed over time.
+
+    This callback can be setup to only log stats on certain EnvRunner indices through
```

Comment:
What happens if an `EnvRunner` crashes?

Reply:
Good question. The index first goes out of commission (other EnvRunners will NOT change their indices b/c of another EnvRunner's crash), but only until the actor is automatically restarted. The latter only happens if `config.recreate_failed_env_runners=True`, of course.